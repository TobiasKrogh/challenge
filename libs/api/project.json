{
  "name": "api",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "libs/api/src",
  "prefix": "api",
  "targets": {
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/{projectRoot}"],
      "options": {
        "jestConfig": "libs/api/jest.config.ts",
        "passWithNoTests": true
      },
      "configurations": {
        "ci": {
          "ci": true,
          "codeCoverage": true
        }
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/api/**/*.ts", "libs/api/**/*.html"]
      }
    }
  },
  "tags": []
}
