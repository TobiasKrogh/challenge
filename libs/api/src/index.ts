export * from './lib/di';
export * from './lib/model';
export * from './lib/services';
