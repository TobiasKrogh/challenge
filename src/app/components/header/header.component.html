<div class="logo">
  <!-- <a href="#"> -->
  <img src="/assets/images/logo.v1.svg" alt="Logo" />
  <!-- </a> -->
</div>

<div class="menu">
  <button class="menu-button" type="button">
    <img src="/assets/images/menu.v1.svg" alt="Menu" />
  </button>
</div>

<ng-container *ngIf="user$ | async as user">
  <challenge-user [user]="user"></challenge-user>
</ng-container>
