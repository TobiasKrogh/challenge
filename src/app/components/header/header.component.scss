:host {
  align-items: center;
  border-bottom: 0.125rem solid var(--hex-color-grey--light);
  display: flex;
  flex: 1;
  height: 5.25rem;
  padding-right: var(--space-right);
}

.logo {
  display: flex;
  justify-content: center;
  width: var(--navigation-width);
}

.menu {
  align-items: center;
  display: flex;
  flex: 1;

  &-button {
    height: 2.6875rem;
    margin-left: -0.375rem;
  }
}
