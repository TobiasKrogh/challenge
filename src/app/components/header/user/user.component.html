<ng-container *ngIf="user">
  <em>
    {{ abbrevation }}
  </em>

  <span>
    {{ fullName }}
  </span>
</ng-container>
