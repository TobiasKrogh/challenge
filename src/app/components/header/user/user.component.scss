:host {
  --full-name-display: none;

  align-items: center;
  display: flex;

  @media (min-width: 480px) {
    --full-name-display: inline;
  }
}

em {
  align-items: center;
  background-color: var(--hex-color-yellow);
  border-radius: var(--border-radius--default);
  color: var(--hex-color-white);
  display: flex;
  font-size: 1.4375rem;
  font-style: normal;
  font-weight: var(--font-weight--bold);
  justify-content: center;

  height: 2.6875rem;
  width: 2.6875rem;
}

span {
  display: var(--full-name-display);
  margin-left: 0.6875rem;
}
