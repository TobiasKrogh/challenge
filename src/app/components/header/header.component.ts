import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { DataService } from '@challenge/api';
import { map } from 'rxjs';

@Component({
  selector: 'challenge-header',
  templateUrl: './header.component.html',
  styleUrls: ['./header.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class HeaderComponent implements OnInit {
  protected readonly user$ = this.dataService.users$.pipe(
    map((users) => users[0])
  );

  constructor(private readonly dataService: DataService) {}

  public ngOnInit(): void {
    this.dataService.loadUsers();
  }
}
