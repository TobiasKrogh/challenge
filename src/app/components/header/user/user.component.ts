import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { IUser } from '@challenge/api';

@Component({
  selector: 'challenge-user',
  templateUrl: './user.component.html',
  styleUrls: ['./user.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class UserComponent {
  @Input() public user: IUser | undefined;

  public get fullName(): string {
    return this.user ? `${this.user.firstName} ${this.user.lastName}` : '';
  }

  public get abbrevation(): string {
    return this.user
      ? `${this.user.firstName.charAt(0)}${this.user.lastName.charAt(0)}`
      : '';
  }
}
