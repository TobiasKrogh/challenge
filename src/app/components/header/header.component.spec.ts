import { ComponentFixture, TestBed } from '@angular/core/testing';
import { DataService } from '@challenge/api';
import { MockComponents } from 'ng-mocks';
import { BehaviorSubject } from 'rxjs';
import { MOCK_USER } from '../../../testing/data.mocks';

import { HeaderComponent } from './header.component';
import { UserComponent } from './user/user.component';

describe('HeaderComponent', () => {
  let component: HeaderComponent;
  let fixture: ComponentFixture<HeaderComponent>;

  let dataService: DataService;
  const users$ = new BehaviorSubject([MOCK_USER]);

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      providers: [
        {
          provide: DataService,
          useValue: {
            loadUsers: jest.fn(),
            users$,
          },
        },
      ],
      declarations: [MockComponents(UserComponent), HeaderComponent],
    }).compileComponents();

    dataService = TestBed.inject(DataService);

    fixture = TestBed.createComponent(HeaderComponent);
    component = fixture.componentInstance;
  });

  it('should create', () => {
    fixture.detectChanges();
    expect(component).toBeTruthy();
  });

  describe('When rendering', () => {
    it('Then it loads users', () => {
      fixture.detectChanges();
      expect(dataService.loadUsers).toHaveBeenCalledWith();
    });

    describe('And user is found', () => {
      it('Then it matches snapshot', () => {
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });
    });

    describe('And no user is found', () => {
      beforeEach(() => {
        users$.next([]);
      });

      it('Then it matches snapshot', () => {
        fixture.detectChanges();
        expect(fixture.debugElement).toMatchSnapshot();
      });
    });
  });
});
