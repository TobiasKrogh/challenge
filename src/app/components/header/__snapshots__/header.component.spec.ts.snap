// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent When rendering And no user is found Then it matches snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root3"
  >
    <div
      class="logo"
    >
      <img
        alt="Logo"
        src="/assets/images/logo.v1.svg"
      />
    </div>
    <div
      class="menu"
    >
      <button
        class="menu-button"
        type="button"
      >
        <img
          alt="Menu"
          src="/assets/images/menu.v1.svg"
        />
      </button>
    </div>
    
  </div>,
}
`;

exports[`HeaderComponent When rendering And user is found Then it matches snapshot 1`] = `
DebugElement {
  "nativeNode": <div
    id="root2"
  >
    <div
      class="logo"
    >
      <img
        alt="Logo"
        src="/assets/images/logo.v1.svg"
      />
    </div>
    <div
      class="menu"
    >
      <button
        class="menu-button"
        type="button"
      >
        <img
          alt="Menu"
          src="/assets/images/menu.v1.svg"
        />
      </button>
    </div>
    <challenge-user />
    
    
  </div>,
}
`;
